gtda/graphs/geodesic_distance.py

```diff
@@ -88,15 +92,29 @@ def __init__(self, n_jobs=None, directed=False, unweighted=False,
         self.unweighted = unweighted
         self.method = method
 
-    def _geodesic_distance(self, X):
-        if not issparse(X) and not isinstance(X, MaskedArray):
-            if X.dtype != bool:
+    def _geodesic_distance(self, X, i=None):
+        method_ = self.method
+        if not issparse(X):
+            diag = np.eye(X.shape[0], dtype=bool)
+            if np.any(~np.logical_or(X, diag)):
+                if self.method in ['auto', 'FW']:
+                    if np.any(X < 0):
+                        method_ = 'J'
+                    else:
+                        method_ = 'D'
+                    warn(
+                        f"Methods 'auto' and 'FW' are not supported when "
+                        f"some edge weights are zero. Using '{method_}' "
+                        f"instead for graph {i}."
+                    )
+            if not isinstance(X, MaskedArray):
                 # Convert to a masked array with mask given by positions in
                 # which infs or NaNs occur.
-                X = masked_invalid(X)
+                if X.dtype != bool:
```

Comment:
Suggestion: aggregate the nested `if`s into one?

Reply:
Oh dear, I tried and developed a headache :) Let me try again.